src/cirrus/lib2/events.py

```diff
@@ -0,0 +1,305 @@
+           function
+        """
+
+        def decorator(function: Callable):
+            @wraps(function)
+            def wrap_function(*args, **kwargs):
+                with cls(*wfem_args, **wfem_kwargs) as wfem:
+                    return function(*args, **kwargs, wfem=wfem)
+
+            return wrap_function
+
+        return decorator
+
+    def announce(
+        self: "WorkflowEventManager",
+        event_type: WFEventType,
+        payload_id: str,
+        payload_url: str = None,
+        extra_message: dict = None,
+    ) -> None:
+        """
+        Construct message payload and publish to WorkflowEventTopic.
+
+        Args:
+            event_type (WFEventType): the type of event which occurred
+            payload_id (str): The ID of the payload.
+            payload_url (str): URL of the ProcessPayload which triggered the event.
+            extra_message (dict): dictionary of additional items to be placed into the
+                                  SNS message body.
+        """
+        if self.event_publisher is None:
+            return
+        if payload_id is None:
+            raise ValueError("Must specify a payload_id")
+
+        message = {
+            "event_type": event_type,
+            "payload_url": payload_url,
+            "payload_id": payload_id,
+        }
+
+        if extra_message:
+            if not all(k not in message for k in extra_message):
+                raise ValueError(
+                    "extra_message parameters must not include: "
+                    + f"{','.join(message.keys())}."
+                )
+            message.update(extra_message)
+        self.event_publisher.add(
+            message=json.dumps(message),
+            message_attrs={
+                "event_type": {"DataType": "String", "StringValue": event_type}
+            },
+        )
+
+    def claim_processing(
+        self: "WorkflowEventManager",
+        payload_id: str,
+        payload_url: str = None,
+        isotimestamp: str = None,
+    ):
+        if not isotimestamp:
+            isotimestamp = datetime.now(timezone.utc).isoformat()
+        self.statedb.claim_processing(payload_id=payload_id, isotimestamp=isotimestamp)
+        self.announce(
+            WFEventType.CLAIMED_PROCESSING,
+            payload_id=payload_id,
+            payload_url=payload_url,
+        )
+
+    def started_processing(
+        self: "WorkflowEventManager",
+        payload_id: str,
+        execution_arn: str,
+        isotimestamp: str = None,
+        payload_url: str = None,
+    ):
+        if not isotimestamp:
+            isotimestamp = datetime.now(timezone.utc).isoformat()
+        self.statedb.set_processing(payload_id, execution_arn, isotimestamp)
+        self._write_timeseries_record(
+            payload_id,
+            state=StateEnum.PROCESSING,
+            event_time=isotimestamp,
+            execution_arn=execution_arn,
+        )
+        self.announce(
```

Comment:
Is it worth including the timestamp in messages where we have one? Seems like that could be valuable, because the message time is not the actual event time.